### DISCUSSION TITLE
Python Typings

### DISCUSSION THREAD

**Discussion by baughmann** [networkit/networkit#979] [Ideas]
Hello all,

Great library, and apologies for the [probably] dumb question in GH issue, but I have not yet been able to find a solution.

I seem unable to import any typings. Are there no Python typings for objects such as `structures.Partition`?

Thanks in advance.

**Comment by fabratu** [comment]
Hi,

thanks for the question! So far, we don't support any Python typings. Out of curiosity - what is the use-case you are trying to solve with it?

In case that already helps, input and output parameters of the API are documented here: https://networkit.github.io/dev-docs/python_api/modules.html

**Comment by baughmann** [comment]
@fabratu  Thanks for your response

Right now, just easy exploration of and toying with the library. In the long run, better project maintainability on my part.

While the documentation is pretty good, I still feel partially lost without some definitions to CTRL+Click to in my IDE. I generally try to type everything I do in Python unless I’m trying to take advantage of dynamic types. Of course, these are personal style decisions.

I think I may be in the minority here in that I’m exploring integrating NetworkIt into an full application rather than simply use it for some one-off analysis, though I may be very wrong about that.

**Comment by fabratu** [comment]
Ah, I see - thank you for the insights! Always interesting to get to know how NetworKit is used. 

Maybe typings is something to consider in the future. However we need to investigate possible usage together with Cython and also have to evaluate possible performance effects (both run-time and loading).

In case that is not already clear, most IDEs also support showing the docstring per shortcut, like Jupyterlab does via `Shift` + `Tab`. For NetworKit the string also includes types for all parameters.

**Reply by baughmann** [reply]
Strange, I don't remember seeing anything pop up in VS Code or Pycharm but will check again when I get home

**Comment by avdgrinten** [comment]
Adding type annotations doesn't have any performance impact. However, I'm not sure if Cython supports them or not.

Many Python projects now seem to encourage their use. I think it would be fine to start adding them to NetworKit.

**Comment by fabratu** [comment]
Ok, good to know concerning the (none-)performance impact and encouraged adoption. Didn't saw this features a lot for other packages until now. Of course I can see the benefits of such an annotation.

Will convert this issue to a discussion, since it is now more about adding a new feature :)
